derp/derp_server.go

```diff
@@ -1452,13 +1470,21 @@ func (c *sclient) sendPong(data [8]byte) error {
 }
 
 // sendPeerGone sends a peerGone frame, without flushing.
-func (c *sclient) sendPeerGone(peer key.NodePublic) error {
-	c.s.peerGoneFrames.Add(1)
+func (c *sclient) sendPeerGone(peer key.NodePublic, reason PeerGoneReasonType) error {
+	switch reason {
+	case PeerGoneReasonDisconnected:
+		c.s.peerGoneDisconnectedFrames.Add(1)
+	case PeerGoneReasonNotHere:
+		c.s.peerGoneNotHereFrames.Add(1)
+	}
 	c.setWriteDeadline()
-	if err := writeFrameHeader(c.bw.bw(), framePeerGone, keyLen); err != nil {
+	data := peer.AppendTo(nil)
```

Comment:
Very slight optimization, but might be worth it in the derp sending path:

```
data := make([]byte, 0, 33) // 32 for key + 1 for reason, there's probably a constant kicking around for this value
data = peer.AppendTo(bs)
data = append(bs, byte(reason))
```

That'll make this send be one alloc instead of two (one for the peer.AppendTo that allocates a 32b slice, then one more to grow the slice for the reason).

Although really this specific send path is infrequently used, so it probably doesn't matter. But in our network-ish code, keeping awareness of allocs-per-packet is a useful background thread to cultivate.